app/src/utils/dependency_graph.rs

```diff
@@ -0,0 +1,140 @@
+    sync::{Arc, LazyLock},
+};
+
+use anyhow::Result;
+use petgraph::{
+    graph::{DiGraph, NodeIndex},
+    visit::{Reversed, Topo},
+    Direction::{Incoming, Outgoing},
+};
+use tokio::sync::Barrier;
+
+use super::ZipLongest;
+use crate::Runtime;
+use comtrya_lib::{
+    actions::Actions::*,
+    contexts::Contexts,
+    manifests::Manifest,
+    utilities::{get_privilege_provider, password_manager::PasswordManager},
+};
+
+type LockedManifest = Arc<LazyLock<Manifest, Box<dyn FnOnce() -> Manifest + Send>>>;
+#[derive(Debug)]
+pub struct DependencyGraph {
+    pub(crate) graph: DiGraph<LockedManifest, ()>,
+    pub(crate) name_to_idx: HashMap<String, NodeIndex>,
+    pub(crate) barrier_map: HashMap<NodeIndex, Arc<Barrier>>,
+}
+
+impl DependencyGraph {
+    pub async fn new(
+        manifests: HashMap<String, Manifest>,
+        contexts: &Contexts,
+        runtime: &mut Runtime,
+    ) -> Result<Self> {
+        let mut this = Self {
+            graph: DiGraph::new(),
+            name_to_idx: HashMap::new(),
+            barrier_map: HashMap::new(),
+        };
+
+        let mut first_package_install: Option<NodeIndex> = None;
+
+        let mut should_ask_for_pass = false;
+        for (_, manifest) in manifests.iter() {
+            let to_index = this.add_manifest(manifest.clone()).await;
+
+            for (dependant, action) in manifest.depends.iter().zip_longest(manifest.actions.iter())
+            {
+                if let Some(dependency_manifest) = dependant.and_then(|dep| manifests.get(dep)) {
+                    let from_index = this.add_manifest(dependency_manifest.clone()).await;
+                    this.graph.add_edge(from_index, to_index, ());
+                }
+
+                should_ask_for_pass = match action {
+                    Some(PackageInstall(_)) | Some(PackageRepository(_)) => {
+                        if let Some(index) = first_package_install {
+                            first_package_install = first_package_install.or(Some(index));
+                            this.graph.add_edge(index, to_index, ());
+                        }
+                        true
+                    }
+                    Some(CommandRun(cva)) => cva.action.privileged,
+                    _ => false,
+                };
+            }
+        }
+
+        this.barrier_map.extend(
+            this.graph
+                .node_indices()
+                .map(|node| {
+                    (
+                        node,
+                        Arc::new(Barrier::new(
+                            this.graph.neighbors_directed(node, Outgoing).count() + 1,
+                        )),
+                    )
+                })
+                .collect::<HashMap<_, _>>(),
+        );
+
+        if should_ask_for_pass {
+            let mut password_manager =
+                PasswordManager::new(get_privilege_provider(contexts).as_deref())?;
+            password_manager.prompt("Please enter password:").await?;
```

Comment:
This is a problem since the implementation in password_manager.rs is for linux only. Is this due to just testing and development on linux or can this functionality not work on windows/macOS/BSD(s)?

Reply:
Due to testing on linux only. I'll need to first get linux working then I can work on windows/macOS/BSD. I just don't know how I'm gonna get it to stop asking for password while running elevated commands.